src/openms/include/OpenMS/ANALYSIS/ID/AhoCorasickDA.h

```diff
@@ -103,142 +88,330 @@ namespace OpenMS
+    UInt32 child_(const UInt32 base_value, const char label);
+    UInt32 child_(const UInt32 base_value, const uint8_t label);
 
-    // calls functions to construct the Aho Corasick
-    void construct_();
+    /// tests if a aa is ambiguous (B, X, J, Z)
+    bool isAmbiguous_(const char s);
+
+
+    //-------------------------------------------------------------------------------------------------------------------------------------
+    //  CDA - retrieval
+    //-------------------------------------------------------------------------------------------------------------------------------------
+
+    //-----------------------------------------------------------------------------------------------------------------
+    // TODO - Question:
+    //  The functions 'failure_' and 'consumeAmbiguousAA_' call 'retrieval_' if a valid transition or spawn was
+    //  found and return false otherwise. If 'retrieval_' returns false, it means that nothing else was found.
+    //  If 'retrieval_' is called in one of the two functions, the result is returned accordingly. But because
+    //  'retrieval_' calls these two functions itself, it is very interlaced. Should the functions only return
+    //  true and in 'retrieval_' even then retrieval_() should be called at true?
+    //  'useSpawn_' too.
+    //-----------------------------------------------------------------------------------------------------------------
 
-    // recursive function to create the trie by subtries
-    void buildTrie_(Size begin, const Size& end, const Size& depth, const UInt32& node_pos);
+    /**
+     * @brief Follows the supply links of a node marked with term_flag to the root and stores all substrings found in multi_seq.
+     */
+    void findSubstrOfHit_();
 
+    /**
+     * @brief checks if a sequence occurs more than once in the given sequences and stores all indices in multi_seq_
+     * @param pep_idx index of found sequence
+     */
+    void multipleOccurrence_(const Size pep_idx );
 
+    /**
+     * @brief
+     * @return returns true if hit was found and false if there is nothing more to found
+     */
     bool retrieval_();
 
-    //expands the BC array and initializes linkage
+    /**
+     * @brief compares the peptide with the end of the sequences stored in Tail
+     * @param tail_str pointer to a given position in Tail
+     * @return true if a sequences was found, otherwise failure is called and returned
+     */
+    bool compareTail_(const Tail_* tail_str);
+
+    /**
+     * @brief set new node_pos if retrieval fails
+     * @return false if no transition was found with the help of the supply link,
+     * if no transition with the next aa was found and no spawn exists.
+     * Otherwise retrieval is called with the new node and peptide position.
+     */
+    bool failure_();
+
+    /**
+     * @brief returns the supply link if this does not eliminate ambiguous aa. Don't do anything special with 0. Must be considered when calling
```

Comment:
this function is only useful for spawns, i.e. when dealing with ambAA's, right?

So should there be a sibling function
`void followSupplyLink(Int32 from, Int32& to)`?

Can you also make input=output, i.e.  merge the first two parameters?